File: src/checkoutservice/internal/database/connection.go
package database

import (
	"context"
	"database/sql"
	"fmt"
	"os"

	secretmanager "cloud.google.com/go/secretmanager/apiv1"
	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
	"github.com/sirupsen/logrus"
	_ "github.com/lib/pq"
)

// Config holds database configuration
type Config struct {
	Host         string
	DatabaseName string
	SecretName   string
	ProjectID    string
}

// Connection represents a database connection
type Connection struct {
	DB  *sql.DB
	log *logrus.Logger
}

// NewConnection creates a new database connection
func NewConnection(log *logrus.Logger) *Connection {
	return &Connection{
		log: log,
	}
}

// Connect initializes the database connection
func (c *Connection) Connect() error {
	config, err := c.loadConfig()
	if err != nil {
		return err
	}

	if config.Host == "" {
		return fmt.Errorf("CLOUDSQL_HOST not set - database connection is required")
	}

	c.log.Info("Initializing Cloud SQL connection for order history...")

	// Get database password from Secret Manager
	password, err := c.getSecretPayload(config.ProjectID, config.SecretName, "latest")
	if err != nil {
		return fmt.Errorf("failed to get database password: %v", err)
	}

	// Create connection string
	dsn := fmt.Sprintf("host=%s user=postgres password=%s dbname=%s sslmode=disable",
		config.Host, password, config.DatabaseName)

	// Open database connection
	db, err := sql.Open("postgres", dsn)
	if err != nil {
		return fmt.Errorf("failed to open database connection: %v", err)
	}

	// Test connection
	if err := db.Ping(); err != nil {
		db.Close()
		return fmt.Errorf("failed to ping database: %v", err)
	}

	c.DB = db
	c.log.Info("Successfully connected to Cloud SQL for order history")

	// Create tables if they don't exist
	if err := c.createTables(); err != nil {
		c.DB.Close()
		c.DB = nil
		return fmt.Errorf("failed to create tables: %v", err)
	}

	return nil
}

// Close closes the database connection
func (c *Connection) Close() error {
	if c.DB != nil {
		return c.DB.Close()
	}
	return nil
}

// loadConfig loads database configuration from environment variables
func (c *Connection) loadConfig() (*Config, error) {
	config := &Config{
		Host:         os.Getenv("CLOUDSQL_HOST"),
		DatabaseName: os.Getenv("ALLOYDB_DATABASE_NAME"),
		SecretName:   os.Getenv("ALLOYDB_SECRET_NAME"),
		ProjectID:    os.Getenv("PROJECT_ID"),
	}

	if config.Host != "" && (config.ProjectID == "" || config.DatabaseName == "" || config.SecretName == "") {
		return nil, fmt.Errorf("missing required environment variables: PROJECT_ID, ALLOYDB_DATABASE_NAME, ALLOYDB_SECRET_NAME")
	}

	return config, nil
}

// getSecretPayload retrieves secret from Google Secret Manager
func (c *Connection) getSecretPayload(projectID, secretID, version string) (string, error) {
	c.log.Infof("Attempting to connect to Secret Manager for project=%s, secret=%s", projectID, secretID)
	
	ctx := context.Background()
	client, err := secretmanager.NewClient(ctx)
	if err != nil {
		c.log.Errorf("Failed to create Secret Manager client: %v", err)
		return "", err
	}
	defer client.Close()

	name := fmt.Sprintf("projects/%s/secrets/%s/versions/%s", projectID, secretID, version)
	c.log.Infof("Accessing secret: %s", name)
	
	req := &secretmanagerpb.AccessSecretVersionRequest{Name: name}

	result, err := client.AccessSecretVersion(ctx, req)
	if err != nil {
		c.log.Errorf("Failed to access secret version: %v", err)
		return "", err
	}

	c.log.Info("Successfully retrieved secret from Secret Manager")
	return string(result.Payload.Data), nil
} 

File: src/reviewservice/database/connection.py
import os
import logging
from contextlib import asynccontextmanager

from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession, async_sessionmaker
from google.cloud import secretmanager

from .models import Base

logger = logging.getLogger(__name__)

class DatabaseManager:
    """Database manager using SQLAlchemy ORM for Cloud SQL PostgreSQL."""
    
    def __init__(self):
        self.engine = None
        self.async_session = None
        self.project_id = os.getenv("PROJECT_ID")
        self.secret_name = os.getenv("ALLOYDB_SECRET_NAME")
        self.database_name = os.getenv("ALLOYDB_DATABASE_NAME", "reviews")
        self.host = os.getenv("CLOUDSQL_HOST")

    async def initialize(self):
        """Initialize the database engine and session factory."""
        if not self.host:
            logger.info("CLOUDSQL_HOST not set - running without database")
            return
        
        logger.info("Initializing Cloud SQL connection for reviews...")
        
        # Get database password from Secret Manager
        password = await self._get_secret_payload(
            self.project_id, 
            self.secret_name, 
            "latest"
        )
        
        # Create async engine
        logger.info(f"Connecting to database URL: postgresql+asyncpg://postgres:***@{self.host}/{self.database_name}")
        database_url = f"postgresql+asyncpg://postgres:{password}@{self.host}/{self.database_name}"
        self.engine = create_async_engine(
            database_url,
            pool_size=5,
            max_overflow=10,
            pool_pre_ping=True,
            echo=True  # Set to True for SQL debugging
        )
        
        # Create session factory
        self.async_session = async_sessionmaker(
            bind=self.engine,
            class_=AsyncSession,
            expire_on_commit=False
        )
        
        logger.info(f"Successfully connected to Cloud SQL for reviews - Database: {self.database_name}, Host: {self.host}")
        
        # Create tables if they don't exist
        await self._create_tables()
    
    async def close(self):
        """Close the database engine."""
        if self.engine:
            await self.engine.dispose()
    
    async def _get_secret_payload(self, project_id: str, secret_id: str, version: str) -> str:
        """Retrieve secret from Google Secret Manager."""
        logger.info(f"Attempting to connect to Secret Manager for project={project_id}, secret={secret_id}")
        
        client = secretmanager.SecretManagerServiceClient()
        name = f"projects/{project_id}/secrets/{secret_id}/versions/{version}"
        
        response = client.access_secret_version(request={"name": name})
        logger.info("Successfully retrieved secret from Secret Manager")
        
        return response.payload.data.decode("UTF-8").strip()
    
    async def _create_tables(self):
        """Create database tables if they don't exist."""
        if not self.engine:
            return
        
        async with self.engine.begin() as conn:
            await conn.run_sync(Base.metadata.create_all)
        
        logger.info("Database tables created/verified")
    
    @asynccontextmanager
    async def get_session(self):
        """Get database session context manager."""
        if not self.async_session:
            raise Exception("Database not initialized")
        
        async with self.async_session() as session:
            try:
                yield session
                await session.commit()
            except Exception:
                await session.rollback()
                raise
            finally:
                await session.close() 